core/jobs/batch_jobs/exp_validation_jobs.py

```diff
@@ -0,0 +1,311 @@
+          interaction having answer group(s) with non-empty training data.
+    """
+
+    def get_states_having_invalid_state_classifier_id(
+        self, states: dict[str, state_domain.State]) -> list[str]:
+        """Returns a list of state names that have not None
+        classifier model id.
+
+        Args:
+            states: dict[str, State]. The mapping of state name to State.
+
+        Returns:
+            list[str]. Returns a list of state names that have not None
+            classifier model id.
+        """
+        invalid_state_names: list[str] = []
+        for state_name in states:
+            if states[state_name].classifier_model_id is not None:
+                invalid_state_names.append(state_name)
+        return invalid_state_names
+
+    def filter_exps_having_invalid_state_classifier_model_id(
+        self, states: dict[str, state_domain.State]) -> bool:
+        """Returns True if any state has classifier model id not None.
+
+        Args:
+            states: dict(str, State). A dict of states.
+
+        Returns:
+            bool. Returns True if any state has classifier model id not None.
+        """
+        return any(
+            state.classifier_model_id is not None for state in states.values()
+        )
+
+    def get_states_having_invalid_training_data(
+        self, exploration: exp_domain.Exploration) -> list[str]:
+        """Returns state names having non-empty training data.
+
+        Args:
+            exploration: Exploration. Exploration to be checked.
+
+        Returns:
+            list[str]. Returns state names having non-empty training data.
+        """
+        state_names: list[str] = []
+        states: dict[str, state_domain.State] = exploration.states
+        for state_name in states:
+            answer_groups: list[state_domain.AnswerGroup] = (
+                states[state_name].interaction.answer_groups)
+            for answer_group in answer_groups:
+                if len(answer_group.training_data) > 0:
+                    state_names.append(state_name)
+                    # The below break is necessary to avoid duplicated
+                    # state names in the list.
+                    break
+        return state_names
+
+    def filter_curated_explorations(
+        self, exp_and_opp_models: tuple[str, dict[str, list[object]]]):
+        """Returns whether the exp model is curated or not.
+
+        Args:
+            exp_and_opp_models: tuple. The pair of exp id and models.
+
+        Returns:
+            bool. Returns whether the exp model is curated or not.
+        """
+        exp_models_list = list(exp_and_opp_models[1]['exp_model'])
+        opp_models_list = list(exp_and_opp_models[1]['exp_opportinity_model'])
+        return len(exp_models_list) == 1 and len(opp_models_list) == 1
+
+    def get_exploration_from_models(
+        self, exp_and_opp_models: tuple[str, dict[str, list[object]]]):
```

Comment:
Again, why object, also missing return type.

Reply:
Modified it now.